File: testdata/precision-suite-csharp/tp_deserialization.cs
// Unsafe deserialization: attacker-controlled bytes are fed to
// BinaryFormatter.Deserialize — CWE-502, a well-known .NET RCE vector. A correct
// scanner fires TAINT-005. BinaryFormatter reconstructs arbitrary object graphs
// and is unsafe on any untrusted input.
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using System.Web;

public class SessionController
{
    // Restore rehydrates a session blob taken straight from the request — the bug.
    public object Restore(HttpRequest Request)
    {
        var blob = Request.Form["state"];
        var formatter = new BinaryFormatter();
        return formatter.Deserialize(ToStream(blob)); // nox-expect: TAINT-005
    }

    private static Stream ToStream(string s) => new MemoryStream(System.Convert.FromBase64String(s));
}

File: testdata/precision-suite-scala/tp_deserialization.scala
// Unsafe deserialization: attacker-controlled request bytes are fed into a native
// ObjectInputStream — CWE-502. A correct scanner fires TAINT-005. Native Java/
// Scala serialization instantiates arbitrary types from the byte stream, a
// well-known remote-code-execution vector; the tainted bytes enter at the stream
// construction.
package controllers

import java.io.ObjectInputStream

class SessionController {
  // restore deserializes untrusted request bytes with no type allow-list — the bug.
  // The stream construction and readObject are a single expression, so the tainted
  // bytes reach the native deserializer in one statement.
  def restore(request: Request): AnyRef = {
    val blob = request.body
    new ObjectInputStream(blob).readObject() // nox-expect: TAINT-005
  }
}
